### PR TITLE
Add APY

### PR DESCRIPTION
**Pull request by jemrickrioux**

### CODE REVIEW

**Comment by tombeynon** [comment]
Amazing work 💥 

I've gone through a few chains, still need to check the rest. Few have issues, hopefully something in common with e.g. Juno and Crypto.org. I'll have a think.

- Osmosis `minting/inflation` not implemented. Same with e-money
- Juno not matching Keplr etc
- Crypto.org seems too low 

General polishing:
- Reset/clear APY when network changes 
- Hide when undefined
- Show in 'Add Validator' modal as it's a clear example of why to choose a restake operator

**Comment by jemrickrioux** [comment]
General polishing:

- [ ] Reset/clear APY when network changes
- [ ] Hide when undefined
- [X] Show in 'Add Validator' modal as it's a clear example of why to choose a restake operator

**Comment by jemrickrioux** [comment]
I added Keplr Suggest Experimental Chain here

**Comment by pwnfoo** [comment]
cheqd denom still looks broken on master, haven't tried your fork yet

image

**Comment by tombeynon** [comment]
@jemrickrioux BRILLIANT knocking out suggest chain too! Will comment on that issue re: extra info but that's excellent.

@pwnfoo I added the decimals for cheqd to `networks.json` in your branch - if you pull that does it solve it? I might separate @jemrickrioux 's work on suggest chain into a different PR so we can get cheqd live using it, as APY might take a little longer. I'll take stock a bit later today

File: src/components/Delegations.js [hunk 1]
```diff
@@ -236,6 +260,13 @@ class Delegations extends React.Component {
     return rewards.reward.find(reward => reward.denom === this.props.network.denom)
   }
 
+  calculateApy(validatorCommission, periodPerYear) {
```
Comment (Sdpierret) [inline src/components/Delegations.js:263]: you need to bonding ratio in the calculation ?

APR = (inflation * (1 - (community tax / 100)) / (bonded tokens / total supply)

but some chain use different calculation (osmosis, juno, evmos)

Comment (jemrickrioux) [reply src/components/Delegations.js:263]: Can you help me with osmosis/Juno/evmos formula please ? (I speak French too)

**Comment by pwnfoo** [comment]
> @pwnfoo I added the decimals for cheqd to `networks.json` in your branch - if you pull that does it solve it? I might separate @jemrickrioux 's work on suggest chain into a different PR so we can get cheqd live using it, as APY might take a little longer. I'll take stock a bit later today

It does add cheqd and I can play around. You can pull it in for now - I won't post a UI for it until the denom work is done :)

File: src/components/Delegations.js (hunk 1 above)
Comment (Sdpierret) [reply src/components/Delegations.js:263]: > Can you help me with osmosis/Juno/evmos formula please ? (I speak French too)

osmosis and evmos are epoch based. You need to evaluate the inflation before dividing by the bonded ratio.
Juno is based on the annual provision and block per year.

I will talk with my team, maybe we can provide a dedicated endpoint.

File: src/components/Validators.js [hunk 2]
```diff
@@ -49,6 +56,7 @@ function Validators(props) {
   function renderItem(item, isOperator){
     let variant = isOperator ? 'warning' : null
     variant = variant ? 'table-' + variant : ''
+    console.log(item)
```
Comment (msteiner96) [inline src/components/Validators.js:59]: Debug? :)

File: src/utils/RestClient.mjs [hunk 3]
```diff
@@ -46,6 +46,28 @@ const RestClient = async (chainId, restUrls) => {
       )
   }
 
+  const getInflation = () => {
```
Comment (msteiner96) [inline src/utils/RestClient.mjs:49]: Let's use cosmjs' `QueryClient` instead.

File: src/components/Validators.js (hunk 2 above)
Comment (jemrickrioux) [reply src/components/Validators.js:59]: Indeed.
I removed it.

Good catch.

**Comment by msteiner96** [comment]
We should probably push https://github.com/eco-stake/restake/pull/88 before this one.

**Comment by jemrickrioux** [comment]
@msteiner I added your changes into my commit to make sure it's easier to merge. I think I solved an issue where validators weren't displaying. I reordered the restClient init params. 

Does that look good? 

Thanks for the help with that

**Comment by jemrickrioux** [comment]
@msteiner96 #88 as been added.
@tombeynon master has been remerged to ease the merging process later on

File: src/components/Delegations.js (hunk 1 above)
Comment (jemrickrioux) [reply src/components/Delegations.js:267]: any chance we can discuss on github about this, I'm in the ecostake discord as mateo | Stake Frites (🥩 , 🍟)#6208

**Comment by jemrickrioux** [comment]
@tombeynon all modifications have been made, I'll now make variations on the calculate apy function to account for other chains (juno, osmo, evmos)

**Comment by jemrickrioux** [comment]
@tombeynon the osmosis thing seems to be an rpc version issue: https://github.com/cosmos/cosmjs/issues/917

**Comment by jemrickrioux** [comment]
@tombeynon

Osmosis formula done
I also corrected the formula for other chains. @Galadrin 

The apy will take the Chain APR, remove the comission and then compound the reward daily. The APY shouldn't match Keplr because de daily compound makes a big difference

To be continued
- [ ] Juno
- [ ] Sifchain
- [ ] Refactor code
- [ ] Use Query client

**Comment by jemrickrioux** [comment]
#215 We ready

**Comment by tombeynon** [comment]
Completed in #215 and #254

**Pull request closed without merging**